### PR TITLE
Add support for SidecarContainers native Kubernetes support

### PR DESCRIPTION
**Pull request by howardjohn**
This finalizes support for native sidecar support (almost). This adds a new feature flag (off by default - its alpha and off in k8s), makes injection support it if so.

The support moves the container from `containers` to `initContainers` when the feature is enabled. It updates a few functions to look in both list of containers as well.

WARNING: this does not actually work e2e *yet*. An update to k8s.io/api needs to be done, and we can't reasonably do that yet. It will be done before Istio 1.19, though.

### CODE REVIEW

File: pkg/kube/inject/webhook.go [hunk 1]
```diff
@@ -662,20 +689,35 @@ func reorderPod(pod *corev1.Pod, req InjectionParameters) error {
...
 
-	// Proxy container should be last, unless HoldApplicationUntilProxyStarts is set
-	// This is to ensure `kubectl exec` and similar commands continue to default to the user's container
-	pod.Spec.Containers = modifyContainers(pod.Spec.Containers, ProxyContainerName, proxyLocation)
-	// Validation container must be first to block any user containers
-	pod.Spec.InitContainers = modifyContainers(pod.Spec.InitContainers, ValidationContainerName, MoveFirst)
-	// Init container must be last to allow any traffic to pass before iptables is setup
-	pod.Spec.InitContainers = modifyContainers(pod.Spec.InitContainers, InitContainerName, MoveLast)
-	pod.Spec.InitContainers = modifyContainers(pod.Spec.InitContainers, EnableCoreDumpName, MoveLast)
+	if features.EnableNativeSidecars.Get() {
```
Comment (howardjohn) [inline pkg/kube/inject/webhook.go:692]: There is one thing majorly broken by this, that we use injection for Gateways and do the rewrite for them.

Options:
1. We somehow indicate per-template whether it is a "sidecar" or not
2. We actually make the template add it as an `initContainer`. this isn't so bad in the template but makes some other logic more complex

I think we should probably do (2)...

**Comment by hzxuzhonghu** [comment]
Just one question: how does startupProbe can solve 1. job exit, sidecar should exit; 2. sidecar run and ready before app container run

And if not, what does it resolve

**Comment by howardjohn** [comment]
startupProbe is just temporary because they didn't implement readinessProbe
yet.

All the things you mentioned are solved in k8s by making it an init
container with RestartPolicy=Always which basically means "it's a sidecar"

**Comment by hanxiaop** [comment]
The changes LGTM, not related to this PR, and I think there may be more UX changes. I'll do a double-check.

**Comment by howardjohn** [comment]
> Just one question: how does startupProbe can solve 1. job exit, sidecar should exit; 2. sidecar run and ready before app container run
> 
> And if not, what does it resolve

https://github.com/kubernetes/kubernetes/pull/119168 is the followup to not need startup probe

**Comment by jacob-delgado** [comment]
Is there an rc yet for us to start testing via kind?

**Comment by howardjohn** [comment]
> Is there an rc yet for us to start testing via kind?

https://github.com/istio/istio/pull/44835 has a build I did - no release tag from upstream yet. You need the feature flag, kind image, and go.mod updates - the 3 things this PR doesn't do

**Review by hanxiaop — APPROVED** [review]

File: pilot/pkg/features/pilot.go [hunk 2]
```diff
@@ -658,6 +658,10 @@ var (
 	KubernetesClientContentType = env.Register("ISTIO_KUBE_CLIENT_CONTENT_TYPE", "protobuf",
 		"The content type to use for Kubernetes clients. Defaults to protobuf. Valid options: [protobuf, json]").Get()
 
+	// This is used in injection templates, it is not unused.
+	EnableNativeSidecars = env.Register("ENABLE_NATIVE_SIDECARS", false,
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/features/pilot.go:662]: No, we can _maybe_ do it after its beta. But in alpha, need explicit opt in.

[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/features/pilot.go:662]: And as you mentioned the many cluster is a concern. We can consider how to set the default once it goes GA

**Review by ericvn — APPROVED** [review]

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**